mlem/contrib/venv.py

```diff
@@ -0,0 +1,228 @@
+import os
+import platform
+import subprocess
+import sys
+import venv
+from abc import abstractmethod
+from types import SimpleNamespace
+from typing import ClassVar, Optional
+
+from mlem.core.errors import MlemError
+from mlem.core.objects import MlemBuilder, MlemModel
+from mlem.ui import EMOJI_OK, EMOJI_PACK, echo
+
+
+class EnvBuilder(MlemBuilder):
+    """MlemBuilder implementation for building virtual environments"""
+
+    type: ClassVar = "env"
+
+    target: Optional[str] = "venv"
+    """Name of the virtual environment"""
+
+    @abstractmethod
+    def create_virtual_env(self):
+        raise NotImplementedError
+
+    @abstractmethod
+    def get_installed_packages(self, env_dir):
+        raise NotImplementedError
+
+    @abstractmethod
+    def build(self, obj: MlemModel) -> str:
+        raise NotImplementedError
+
+
+class VenvBuilder(EnvBuilder):
+    class Config:
+        arbitrary_types_allowed = True
+
+    type: ClassVar = "venv"
+
+    no_cache: Optional[bool] = False
+    """Disable cache"""
+    current_env: Optional[bool] = False
+    """Whether to install in the current virtual env, must be active"""
+    context: Optional[SimpleNamespace] = None
+    """context for the virtual env"""
+
+    def create_virtual_env(self):
+        env_spec = venv.EnvBuilder(with_pip=True)
+        env_dir = os.path.abspath(self.target)
+
+        self.context = env_spec.ensure_directories(env_dir)
+
+        true_system_site_packages = env_spec.system_site_packages
+        env_spec.system_site_packages = False
+        env_spec.create_configuration(self.context)
+        env_spec.setup_python(self.context)
+        if env_spec.with_pip:
+            env_spec._setup_pip(  # pylint: disable=protected-access
```

Comment:
It seems you always create spec with `with_pip=True`. And why you need protected access?

Reply:
because `_setup_pip` is a protected function.
And we need pip in the new environment so as to use it to install packages with it.

Reply:
Yes, but it is protected for some reason. That means you are not using it how it is expected. You should investigate why is it protected and either find another way to use it (maybe there is a public wrapper or something) or acknowledge that you are "trespassing" and have a good reason for that as well as understanding how it can break and be ready for this

Reply:
removed usage of `context` altogether so we don't have the issue of `protected access` now